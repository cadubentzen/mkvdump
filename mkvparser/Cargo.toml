[package]
name = "mkvparser"
version = "0.4.4"
authors = ["Carlos Bentzen <cadubentzen@gmail.com>"]
edition = "2021"
license = "MIT OR Apache-2.0"
description = "MKV and WebM parser"
repository = "https://github.com/cadubentzen/mkvdump"
keywords = ["mkv", "matroska", "webm", "ebml"]
categories = ["multimedia", "parser-implementations"]
exclude = ["inputs/*", "tests/snapshots/*", ".github/*"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
nom = "7"
serde = { version = "1.0", features = ["derive"] }
chrono = { version = "0.4", default-features = false, features = ["serde"] }
serde_with = "3"
thiserror = "1"

[build-dependencies]
serde = { version = "1.0", features = ["derive"] }
serde-xml-rs = "0.6"
convert_case = "0.6"

[dev-dependencies]
serde_yaml = "0.9"
