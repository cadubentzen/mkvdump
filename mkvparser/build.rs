use std::fs::File;
use std::io::Write;
use std::path::Path;
use std::{collections::HashSet, env};

use convert_case::{Case, Casing};
use serde::{Deserialize, Serialize};

const EBML_XML: &str = include_str!("ebml.xml");
const EBML_MATROSKA_XML: &str = include_str!("ebml_matroska.xml");

#[derive(Debug, PartialEq, Serialize, Deserialize)]
struct EBMLSchema {
    #[serde(rename(deserialize = "$value"))]
    elements: Vec<Element>,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
struct Element {
    name: String,
    path: String,
    id: String,
    #[serde(rename(deserialize = "type"))]
    variant: String,
    #[serde(rename(deserialize = "$value"))]
    details: Option<Vec<ElementDetail>>,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
enum ElementDetail {
    Documentation(Documentation),
    Extension(Extension),
    Restriction(Restriction),
    ImplementationNote(ImplementationNote),
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
struct Documentation {
    #[serde(rename(deserialize = "$value"))]
    text: String,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
struct Extension {
    webm: Option<bool>,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
struct Restriction {
    #[serde(rename(deserialize = "$value"))]
    enums: Vec<Enum>,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
struct Enum {
    value: String,
    label: String,
    #[serde(rename(deserialize = "$value"))]
    documentation: Option<Vec<Documentation>>,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
struct ImplementationNote {
    #[serde(rename(deserialize = "$value"))]
    text: String,
}

fn get_elements() -> Vec<Element> {
    let ebml_schema: EBMLSchema = serde_xml_rs::from_str(EBML_XML).unwrap();
    let ebml_matroska_schema: EBMLSchema = serde_xml_rs::from_str(EBML_MATROSKA_XML).unwrap();

    // Ignoring Matroska overrides of EBML elements
    let mut known_elements = HashSet::<String>::new();
    let mut elements = Vec::<Element>::new();
    for element in ebml_schema
        .elements
        .into_iter()
        .chain(ebml_matroska_schema.elements.into_iter())
    {
        if known_elements.get(&element.name).is_none() {
            known_elements.insert(element.name.clone());
            elements.push(element);
        }
    }

    // Pre-format names and variants
    elements.iter_mut().for_each(|e| {
        e.variant = variant_to_enum_literal(&e.variant).to_string();
    });

    elements
}

fn variant_to_enum_literal(variant: &str) -> &str {
    match variant {
        "master" => "Master",
        "uinteger" => "Unsigned",
        "integer" => "Signed",
        "string" => "String",
        "binary" => "Binary",
        "utf-8" => "Utf8",
        "date" => "Date",
        "float" => "Float",
        _ => panic!("Variant not expected: {}", variant),
    }
}

fn apply_label_quirks(label: &str, reserved_index: &mut i32) -> String {
    let mut label = label
        .replace(|c: char| !c.is_ascii_alphanumeric(), " ")
        .to_case(Case::Pascal);

    // Hack because identifiers can't start with a number
    if label == "3Des" {
        label = "TripleDes".to_string();
    }
    // "Reserved" sometimes repeats in enums
    else if label == "Reserved" {
        label = format!("Reserved{}", reserved_index);
        *reserved_index += 1;
    }

    label
}

fn create_elements_file(elements: &[Element]) -> std::io::Result<()> {
    let out_dir = env::var_os("OUT_DIR").unwrap();
    let elements_path = Path::new(&out_dir).join("elements.rs");
    let mut file = File::create(elements_path)?;

    writeln!(file, "use crate::ebml::ebml_elements;")?;
    writeln!(file, "ebml_elements! {{")?;

    for Element {
        name,
        id,
        variant,
        path: _,
        details,
    } in elements
    {
        if let Some(details) = details {
            macro_rules! write_comment_lines {
                ($detail_type:path) => {
                    for detail in details {
                        if let $detail_type(detail) = detail {
                            detail.text.split('\n').filter(|s| !s.is_empty()).for_each(
                                |doc_line| writeln!(file, "    /// {}", doc_line).unwrap(),
                            );
                        }
                    }
                };
            }

            write_comment_lines!(ElementDetail::Documentation);
            write_comment_lines!(ElementDetail::ImplementationNote);
        }

        let enum_name = name.to_case(Case::Pascal);
        writeln!(
            file,
            "    name = {enum_name}, original_name = \"{name}\", id = {id}, variant = {variant};"
        )?;
    }
    writeln!(file, "}}")?;

    Ok(())
}

fn create_enumerations_file(elements: &[Element]) -> std::io::Result<()> {
    let out_dir = env::var_os("OUT_DIR").unwrap();
    let enumerations_path = Path::new(&out_dir).join("enumerations.rs");
    let mut file = File::create(enumerations_path)?;

    writeln!(file, "use crate::ebml::ebml_enumerations;")?;
    writeln!(file, "ebml_enumerations! {{")?;

    for element in elements {
        let mut reserved_index = 1;
        if element.variant != "Unsigned" {
            continue;
        }
        let enum_name = element.name.to_case(Case::Pascal);
        if let Some(details) = &element.details {
            let mut documentation = None;
            for detail in details {
                match detail {
                    ElementDetail::Documentation(doc) => {
                        documentation = Some(doc.text.clone());
                    }
                    ElementDetail::Restriction(restriction) => {
                        if let Some(ref doc) = documentation {
                            doc.split('\n')
                                .filter(|s| !s.is_empty())
                                .for_each(|doc_line| {
                                    writeln!(file, "    /// {}", doc_line).unwrap();
                                });
                        }
                        writeln!(file, "    {} {{", enum_name)?;
                        for enumeration in &restriction.enums {
                            if let Some(docs) = &enumeration.documentation {
                                for doc in docs {
                                    doc.text.split('\n').filter(|s| !s.is_empty()).for_each(
                                        |doc_line| {
                                            writeln!(file, "        /// {}", doc_line).unwrap()
                                        },
                                    );
                                }
                            } else {
                                writeln!(file, "        /// {}", enumeration.label)?;
                            }
                            let label = apply_label_quirks(&enumeration.label, &mut reserved_index);
                            writeln!(
                                file,
                                "        {} = {}, original_label = \"{}\";",
                                label, enumeration.value, enumeration.label
                            )?;
                        }
                        writeln!(file, "    }};")?;
                    }
                    _ => (),
                }
            }
        }
    }
    writeln!(file, "}}")?;

    Ok(())
}

fn main() -> std::io::Result<()> {
    println!("cargo:rerun-if-changed=ebml.xml");
    println!("cargo:rerun-if-changed=ebml_matroska.xml");

    let elements = get_elements();
    create_elements_file(&elements)?;
    create_enumerations_file(&elements)?;

    Ok(())
}
