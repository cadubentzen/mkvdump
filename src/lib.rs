use std::{fs::File, io::Read, path::Path};

use anyhow::bail;
use mkvparser::{find_valid_element, parse_element, Body, Element};

fn insert_position(element: &mut Element, position: &mut Option<usize>) {
    element.header.position = *position;
    *position = position.map(|p| {
        if let Body::Master = element.body {
            p + element.header.header_size
        } else {
            // It's safe to unwrap because all non-Master elements have a set size
            p + element.header.size.unwrap()
        }
    });
}

#[doc(hidden)]
pub fn parse_elements_from_file(
    path: impl AsRef<Path>,
    show_positions: bool,
    buffer_size: u64,
) -> anyhow::Result<Vec<Element>> {
    let mut file = File::open(path)?;
    let file_length = file.metadata()?.len();

    let buffer_size = file_length.min(buffer_size).try_into().unwrap();
    let mut buffer = vec![0; buffer_size];
    let mut filled = 0;
    let mut elements = Vec::<Element>::new();
    let mut position = show_positions.then_some(0);

    loop {
        let num_read = file.read(&mut buffer[filled..])?;
        if num_read == 0 {
            if filled == 0 {
                // we have nothing left to read or parse
                break;
            } else if filled == buffer_size {
                // We might arrive at this condition for two reasons:
                // - next element does not fit in the buffer, so we continuously get NeedData
                // - buffer is completely filled with corrupt data
                bail!("failed to parse the given file with buffer size of {buffer_size} bytes");
            }
        }

        let mut parse_buffer = &buffer[..(filled + num_read)];
        loop {
            match parse_element(&parse_buffer) {
                Ok((new_parse_buffer, mut element)) => {
                    println!("{element:?}");
                    insert_position(&mut element, &mut position);
                    elements.push(element);
                    parse_buffer = new_parse_buffer;
                }
                Err(e) => {
                    match e {
                        mkvparser::Error::NeedData => break,
                        _ => {
                            match find_valid_element(parse_buffer) {
                                Ok((new_parse_buffer, mut corrupt_element)) => {
                                    // TODO: merge consecutive corrupt elements
                                    insert_position(&mut corrupt_element, &mut position);
                                    elements.push(corrupt_element);
                                    parse_buffer = new_parse_buffer;
                                }
                                Err(_) => {
                                    // TODO: if we can't find a valid element, we could consume the
                                    // whole buffer as a corrupt element. That way we wouldn't bail if
                                    // the corrupt region is bigger than the buffer size.
                                    break;
                                }
                            }
                        }
                    }
                    break;
                }
            }
        }
        filled = parse_buffer.len();
        let parse_buffer = Vec::from(parse_buffer);
        buffer[..filled].copy_from_slice(&parse_buffer);
    }
    Ok(elements)
}
