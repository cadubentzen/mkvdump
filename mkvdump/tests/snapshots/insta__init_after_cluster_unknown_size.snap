---
source: src/lib.rs
assertion_line: 1039
expression: "parse_buffer_to_end(include_bytes!(\"../inputs/init_after_cluster_unknown_size.webm\"))"
---
- id: Cluster
  header_size: 6
  size: Unknown
  children:
    - id: Timestamp
      header_size: 2
      size: 4
      value: 1084
    - id: SimpleBlock
      header_size: 3
      size: 16068
      value:
        track_number: 1
        timestamp: 0
        keyframe: true
    - id: SimpleBlock
      header_size: 3
      size: 7179
      value:
        track_number: 1
        timestamp: 42
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 84
    - id: SimpleBlock
      header_size: 3
      size: 169
      value:
        track_number: 1
        timestamp: 126
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 167
    - id: SimpleBlock
      header_size: 3
      size: 1995
      value:
        track_number: 1
        timestamp: 209
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 251
    - id: SimpleBlock
      header_size: 3
      size: 517
      value:
        track_number: 1
        timestamp: 292
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 334
    - id: SimpleBlock
      header_size: 3
      size: 1597
      value:
        track_number: 1
        timestamp: 376
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 418
    - id: SimpleBlock
      header_size: 3
      size: 753
      value:
        track_number: 1
        timestamp: 459
    - id: SimpleBlock
      header_size: 2
      size: 35
      value:
        track_number: 1
        timestamp: 501
    - id: SimpleBlock
      header_size: 3
      size: 14709
      value:
        track_number: 1
        timestamp: 543
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 584
    - id: SimpleBlock
      header_size: 3
      size: 1173
      value:
        track_number: 1
        timestamp: 626
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 668
    - id: SimpleBlock
      header_size: 3
      size: 3421
      value:
        track_number: 1
        timestamp: 709
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 751
    - id: SimpleBlock
      header_size: 3
      size: 1027
      value:
        track_number: 1
        timestamp: 793
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 835
    - id: SimpleBlock
      header_size: 3
      size: 1822
      value:
        track_number: 1
        timestamp: 876
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 918
    - id: SimpleBlock
      header_size: 3
      size: 981
      value:
        track_number: 1
        timestamp: 960
    - id: SimpleBlock
      header_size: 2
      size: 32
      value:
        track_number: 1
        timestamp: 1001
    - id: SimpleBlock
      header_size: 4
      size: 16876
      value:
        track_number: 1
        timestamp: 1043
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 1085
    - id: SimpleBlock
      header_size: 3
      size: 862
      value:
        track_number: 1
        timestamp: 1127
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 1168
    - id: SimpleBlock
      header_size: 3
      size: 2715
      value:
        track_number: 1
        timestamp: 1210
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 1252
    - id: SimpleBlock
      header_size: 3
      size: 755
      value:
        track_number: 1
        timestamp: 1293
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 1335
    - id: SimpleBlock
      header_size: 3
      size: 1510
      value:
        track_number: 1
        timestamp: 1377
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 1419
    - id: SimpleBlock
      header_size: 3
      size: 722
      value:
        track_number: 1
        timestamp: 1460
    - id: SimpleBlock
      header_size: 2
      size: 33
      value:
        track_number: 1
        timestamp: 1502
    - id: SimpleBlock
      header_size: 4
      size: 19601
      value:
        track_number: 1
        timestamp: 1544
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 1585
    - id: SimpleBlock
      header_size: 3
      size: 460
      value:
        track_number: 1
        timestamp: 1627
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 1669
    - id: SimpleBlock
      header_size: 3
      size: 1699
      value:
        track_number: 1
        timestamp: 1710
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 1752
    - id: SimpleBlock
      header_size: 3
      size: 359
      value:
        track_number: 1
        timestamp: 1794
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 1836
    - id: SimpleBlock
      header_size: 3
      size: 746
      value:
        track_number: 1
        timestamp: 1877
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 1919
    - id: SimpleBlock
      header_size: 3
      size: 307
      value:
        track_number: 1
        timestamp: 1961
    - id: SimpleBlock
      header_size: 2
      size: 32
      value:
        track_number: 1
        timestamp: 2002
    - id: SimpleBlock
      header_size: 3
      size: 10673
      value:
        track_number: 1
        timestamp: 2044
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 2086
    - id: SimpleBlock
      header_size: 3
      size: 229
      value:
        track_number: 1
        timestamp: 2128
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 2169
    - id: SimpleBlock
      header_size: 3
      size: 1181
      value:
        track_number: 1
        timestamp: 2211
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 2253
    - id: SimpleBlock
      header_size: 3
      size: 214
      value:
        track_number: 1
        timestamp: 2294
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 2336
    - id: SimpleBlock
      header_size: 3
      size: 564
      value:
        track_number: 1
        timestamp: 2378
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 2420
    - id: SimpleBlock
      header_size: 3
      size: 214
      value:
        track_number: 1
        timestamp: 2461
    - id: SimpleBlock
      header_size: 2
      size: 32
      value:
        track_number: 1
        timestamp: 2503
    - id: SimpleBlock
      header_size: 3
      size: 7248
      value:
        track_number: 1
        timestamp: 2545
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 2586
    - id: SimpleBlock
      header_size: 3
      size: 183
      value:
        track_number: 1
        timestamp: 2628
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 2670
    - id: SimpleBlock
      header_size: 3
      size: 557
      value:
        track_number: 1
        timestamp: 2711
    - id: SimpleBlock
      header_size: 2
      size: 9
      value:
        track_number: 1
        timestamp: 2753
    - id: SimpleBlock
      header_size: 3
      size: 252
      value:
        track_number: 1
        timestamp: 2795
    - id: SimpleBlock
      header_size: 3
      size: 170
      value:
        track_number: 1
        timestamp: 2837
    - id: SimpleBlock
      header_size: 2
      size: 33
      value:
        track_number: 1
        timestamp: 2878
- id: EBML
  header_size: 5
  size: 36
  children:
    - id: EBMLVersion
      header_size: 3
      size: 4
      value: 1
    - id: EBMLReadVersion
      header_size: 3
      size: 4
      value: 1
    - id: EBMLMaxIDLength
      header_size: 3
      size: 4
      value: 4
    - id: EBMLMaxSizeLength
      header_size: 3
      size: 4
      value: 8
    - id: DocType
      header_size: 3
      size: 7
      value: webm
    - id: DocTypeVersion
      header_size: 3
      size: 4
      value: 2
    - id: DocTypeReadVersion
      header_size: 3
      size: 4
      value: 2
- id: Segment
  header_size: 12
  size: Unknown
  children:
    - id: Void
      header_size: 9
      size: 229
      value: ~
    - id: Info
      header_size: 5
      size: 44
      children:
        - id: TimestampScale
          header_size: 4
          size: 7
          value: 1000000
        - id: MuxingApp
          header_size: 3
          size: 16
          value: Lavf58.29.100
        - id: WritingApp
          header_size: 3
          size: 16
          value: Lavf58.29.100
    - id: Tracks
      header_size: 5
      size: 101
      children:
        - id: TrackEntry
          header_size: 9
          size: 96
          children:
            - id: TrackNumber
              header_size: 2
              size: 3
              value: 1
            - id: TrackUID
              header_size: 3
              size: 4
              value: 1
            - id: FlagLacing
              header_size: 2
              size: 3
              value: 0
            - id: Language
              header_size: 4
              size: 7
              value: und
            - id: CodecID
              header_size: 2
              size: 7
              value: V_AV1
            - id: TrackType
              header_size: 2
              size: 3
              value: video
            - id: DefaultDuration
              header_size: 4
              size: 8
              value: 41708333
            - id: Video
              header_size: 9
              size: 32
              children:
                - id: PixelWidth
                  header_size: 2
                  size: 4
                  value: 1280
                - id: PixelHeight
                  header_size: 2
                  size: 4
                  value: 720
                - id: Colour
                  header_size: 3
                  size: 15
                  children:
                    - id: Range
                      header_size: 3
                      size: 4
                      value: broadcast range
                    - id: ChromaSitingHorz
                      header_size: 3
                      size: 4
                      value: left collocated
                    - id: ChromaSitingVert
                      header_size: 3
                      size: 4
                      value: half
            - id: CodecPrivate
              header_size: 3
              size: 20
              value: "[81 05 0c 00 0a 0b 00 00 00 2d 4c ff b3 df ff 98 04]"
    - id: Tags
      header_size: 5
      size: 61
      children:
        - id: Tag
          header_size: 10
          size: 56
          children:
            - id: Targets
              header_size: 10
              size: 10
              children: []
            - id: SimpleTag
              header_size: 10
              size: 36
              children:
                - id: TagName
                  header_size: 3
                  size: 10
                  value: ENCODER
                - id: TagString
                  header_size: 3
                  size: 16
                  value: Lavf58.29.100

